probe-rs/src/architecture/arm/sequences/infineon.rs

```diff
@@ -0,0 +1,249 @@
+        } else if fp_ctrl.rev() == 1 {
+            FpRev2CompX::breakpoint_configuration(application_entry).into()
+        } else {
+            return Err(Error::Other(anyhow!(
+                "xmc4000: unexpected fp_ctrl.rev = {}",
+                fp_ctrl.rev()
+            )));
+        };
+        core.write_word_32(FpRev1CompX::ADDRESS, val)?;
+        tracing::debug!("Set a breakpoint at {:08x}", application_entry);
+
+        core.flush()?;
+
+        Ok(())
+    }
+
+    fn reset_catch_clear(
+        &self,
+        core: &mut Memory,
+        _core_type: probe_rs_target::CoreType,
+        _debug_base: Option<u64>,
+    ) -> Result<(), Error> {
+        tracing::trace!("performing XMC4000 ResetCatchClear");
+
+        // Grab the prior breakpoint state
+        let (enabled, previous_breakpoint) = self
+            .halt_after_reset_state
+            .lock()
+            .map(|mut m| m.take().unwrap_or((false, 0)))
+            .unwrap();
+
+        // Put FPCTRL back
+        let mut fpctrl = FpCtrl::from(0);
+        fpctrl.set_key(true);
+        fpctrl.set_enable(enabled);
+        core.write_word_32(FpCtrl::ADDRESS, fpctrl.into())?;
+
+        // Put FPCOMP0 back
+        core.write_word_32(FpRev1CompX::ADDRESS, previous_breakpoint)?;
+
+        Ok(())
+    }
+
+    fn reset_system(
+        &self,
+        core: &mut Memory,
+        _core_type: probe_rs_target::CoreType,
+        _debug_base: Option<u64>,
+    ) -> Result<(), Error> {
+        // XMC4700/XMC4800 reference manual v1.3 § 27.2.2.2:
+        // > Since the Reset Status Information in register SCU.RSTSTAT is the accumulated reset
+        // > type, it is necessary to clean the bitfield using the SCU register RSTCLR.RSCLR before
+        // > issuing a System Reset, else SSW will enter the boot mode reflected in STCON.HWCON.
+        //
+        // ARM KB https://developer.arm.com/documentation/ka002435:
+        // > After the user program has handled the the reset sources itself, it needs to clear them
+        // > by executing:
+        // >
+        // >    SCU_RESET->RSTCLR = SCU_RESET_RSTCLR_RSCLR_Msk ;
+        // >
+        // > before the next reset. This will clear also the PORST bit. And if the next reset is no
+        // > Power On Reset, the SSW will also not do HAR.
+        //
+        // This is normally handled by the runtime, but let's be defensive.
+        //
+        // SCU_RESET->RSTCLR is at 0x5000_4408, and RSCLR is the low bit.
+        core.write_word_32(0x5000_4408, 1)?;
+        tracing::debug!("Cleared SCU_RESET->RSTSTAT");
+
+        let mut aircr = Aircr(0);
+        aircr.vectkey();
+        aircr.set_sysresetreq(true);
+        core.write_word_32(Aircr::ADDRESS, aircr.into())?;
+        tracing::debug!("Resetting via AIRCR.SYSRESETREQ");
+
+        // Do not drive the debug pins for a short while
+        // We do not want to clobber the boot mode
+        std::thread::sleep(Duration::from_millis(100));
```

Comment:
Is this always enough? Is this documented or just a wild guess? We could make it configurable if need be.

Reply:
🤔 That's a great question! I'm not sure.

I added it when I thought that section might be setting `HWCON`, but now I understand that `SYSRESET` cannot affect `HWCON` anyway. I now think it might be papering over the period where the system software prohibits debug access:

<img width="508" alt="XMC4700 / XMC4800 documentation, reading: For security reasons it is required to prevent a debug access to the processor before and while the boot firmware code from ROM (SSW) is being executed. A bit DAPSA, (DAP has system access) in the SCU is implemented, allowing the access from CoreSightTM debug system to the processor core. The default value of this bit is disabled debug access. The register is reset by System Reset. The System Reset disables the debug access each time SSW is being executed. At the end of the SSW the DAPSA is enabled always (independent of any other register setting or signaling value), to allow debug access to the CPU. A tool accessing the SoC during the SSW execution time reads back a zero and a write is going to a virtual, none existing address." src="https://user-images.githubusercontent.com/118362/205450292-d9657eb8-2bc9-4841-bc2d-0d769a1efa0b.png">

I'll remove it and see what happens.

Reply:
I removed it with no ill effects. I did however add another fixed 100ms timer in the new `reset_hardware_assert()` sequence 😆

There doesn't seem to be any configurability with the time between `reset_hardware_assert()` and `reset_hardware_deassert()` elsewhere. 100ms is a million times longer than the 100ns guaranteed to cause a reset on this platform, so it ought to be fine.